lib/puma/server.rb

```diff
@@ -583,32 +583,35 @@ def response_to_error(client, requests, err, status_code)
     # Wait for all outstanding requests to finish.
     #
     def graceful_shutdown
-      if options[:shutdown_debug]
-        threads = Thread.list
-        total = threads.size
+      shutdown_debugger = if options[:shutdown_debug]
+        lambda do |message|
+          threads = Thread.list
+          total = threads.size
 
-        pid = Process.pid
+          pid = Process.pid
 
-        $stdout.syswrite "#{pid}: === Begin thread backtrace dump ===\n"
+          $stdout.syswrite "#{pid}: #{message}\n"
+          $stdout.syswrite "#{pid}: === Begin thread backtrace dump ===\n"
 
-        threads.each_with_index do |t,i|
-          $stdout.syswrite "#{pid}: Thread #{i+1}/#{total}: #{t.inspect}\n"
-          $stdout.syswrite "#{pid}: #{t.backtrace.join("\n#{pid}: ")}\n\n"
+          threads.each_with_index do |t,i|
+            $stdout.syswrite "#{pid}: Thread #{i+1}/#{total}: #{t.inspect}\n"
+            $stdout.syswrite "#{pid}: #{t.backtrace.join("\n#{pid}: ")}\n\n"
+          end
+
+          $stdout.syswrite "#{pid}: === End thread backtrace dump ===\n"
         end
-        $stdout.syswrite "#{pid}: === End thread backtrace dump ===\n"
+      end
+
+      if shutdown_debugger && options[:shutdown_debug] != :on_force
+        shutdown_debugger.call("Shutdown initiated")
```

Comment:
I considered moving this call to the `ThreadPool#shutdown` method as well, so we don't have to pass a proc around. But that would mean a change in behaviour where we print after closing the binder instead of before. I don't know if this really matters? If not, moving this would make the implementation a lot cleaner.

Reply:
Sweet, I've refactored, which was quite straightforward now that there's coverage.